keys/src/lib.rs

```diff
@@ -0,0 +1,193 @@
+extern crate ethsign;
+extern crate libp2p;
+extern crate rand;
+
+use ethsign::{Protected, SecretKey};
+use ethsign::keyfile::{Bytes, KeyFile};
+use libp2p::secio::SecioKeyPair;
+use std::error::Error;
+use std::io::prelude::*;
+use std::fs::File;
+use std::num::NonZeroU32;
+use std::path::Path;
+
+pub fn generate_key() -> Result<SecioKeyPair, Box<dyn Error + Send + Sync>> {
+    SecioKeyPair::secp256k1_generated()
+}
+
+pub fn generate_raw() -> [u8; 32] {
+    rand::random()
+}
+
+pub fn raw_to_key(raw_key: [u8; 32]) -> 
+    Result<SecioKeyPair, Box<dyn Error + Send + Sync>> 
+{
+    SecioKeyPair::secp256k1_raw_key(raw_key)
+}
+
+pub fn save_raw_key<P>(key: [u8; 32], file_path: P) ->
+    Result<usize, Box<dyn Error + Send + Sync>>
+where
+    P: AsRef<Path>
+{
+    let mut file = File::create(file_path)?;
+    let res = file.write(&key)?;
+    Ok(res)
+}
+
+
+pub fn save_wallet<P>(key: [u8; 32], file_path: P, password: String) -> 
+    Result<(), Box<dyn Error + Send + Sync>>
+where
+    P: AsRef<Path>
+{
+    let secret = SecretKey::from_raw(&key)?;
+    let crypto = secret.to_crypto(&Protected::new(password.as_bytes().to_vec()), 
+                                  NonZeroU32::new(4096).unwrap())?;
+    let keyfile = KeyFile {
+        id: "".into(),
+        version: 3,
+        crypto: crypto,
+        address: Some(Bytes(secret.public().address().to_vec()))
+    };
+    let keyfile_str = serde_json::to_string(&keyfile)?;
+    let mut file = File::create(file_path)?;
+    let res = file.write_all(keyfile_str.as_bytes())?;
+    Ok(res)
+}
+
+
+pub fn load_key<P>(file_path: P) -> 
+    Result<SecioKeyPair, Box<dyn Error + Send + Sync>>
+where
+    P: AsRef<Path>
+{
+    let mut file = File::open(file_path)?;
+    let mut raw_key = [0; 32];
+    file.read(&mut raw_key)?;
+    SecioKeyPair::secp256k1_raw_key(raw_key)
+}
+
+pub fn load_wallet<P>(file_path: P, password: String) -> 
+    Result<SecioKeyPair, Box<dyn Error + Send + Sync>>
+where
+    P: AsRef<Path>
+{
+    let mut file = File::open(file_path)?;
+    let mut wallet = String::new();
+    file.read_to_string(&mut wallet)?;
+    let keyfile: KeyFile = serde_json::from_str(&wallet.as_str())?;
+    let plain = keyfile.crypto.decrypt(
+        &Protected::new(password.as_bytes().to_vec()))?;
+    SecioKeyPair::secp256k1_raw_key(plain)
+}
+
+
+pub fn load_or_generate_wallet<P>(file_path: &P, password: String) -> 
+    Result<SecioKeyPair,Box<dyn Error + Send + Sync>> 
+where
+    P: AsRef<Path>
+{
+    let res = load_wallet(file_path, password.clone());
+    match res {
+        Ok(_) => res,
+        Err(_) => {
```

Comment:
I don't really like this approach. Loading a wallet can fail for other reasons than the file not being there, right? Imagine someone provided a wrong password by accident. We would overwrite his wallet in such case. So maybe make an explicit check if the wallet file is present and generate the key otherwise.

Reply:
Check now - it's not perfect solution, but resolves main problem. Anyway this function should be rather used for our internal tests right now.